File: test files/java/pleno_iniciante.java
import java.time.Clock;
import java.util.Optional;

public class PlenoIniciante {
    enum TaskStatus {
        TODO, DOING, DONE, BLOCKED
    }

    record Task(int id, String title, Integer ownerId, TaskStatus status, int estimateHours) {}

    record TaskCompletionResult(boolean ok, String code, Task task) {}

    record AuditEntry(int taskId, String action, String at) {}

    interface TaskRepository {
        Optional<Task> findById(int id);
        Task save(Task task);
    }

    interface AuditGateway {
        void publish(AuditEntry entry);
    }

    static class TaskPolicy {
        Optional<String> validateForCompletion(Task task) {
            if (task.ownerId() == null) {
                return Optional.of("TASK_WITHOUT_OWNER");
            }

            if (task.status() == TaskStatus.BLOCKED) {
                return Optional.of("TASK_BLOCKED");
            }

            if (task.estimateHours() <= 0) {
                return Optional.of("INVALID_ESTIMATE");
            }

            return Optional.empty();
        }
    }

    static class TaskWorkflowService {
        private final TaskRepository taskRepository;
        private final AuditGateway auditGateway;
        private final TaskPolicy taskPolicy;
        private final Clock clock;

        TaskWorkflowService(TaskRepository taskRepository, AuditGateway auditGateway, TaskPolicy taskPolicy, Clock clock) {
            this.taskRepository = taskRepository;
            this.auditGateway = auditGateway;
            this.taskPolicy = taskPolicy;
            this.clock = clock;
        }

        TaskCompletionResult moveToDone(int taskId) {
            Optional<Task> maybeTask = taskRepository.findById(taskId);
            if (maybeTask.isEmpty()) {
                return new TaskCompletionResult(false, "TASK_NOT_FOUND", null);
            }

            Task task = maybeTask.get();
            Optional<String> validationError = taskPolicy.validateForCompletion(task);
            if (validationError.isPresent()) {
                return new TaskCompletionResult(false, validationError.get(), task);
            }

            Task updatedTask = new Task(
                task.id(),
                task.title(),
                task.ownerId(),
                TaskStatus.DONE,
                task.estimateHours()
            );

            taskRepository.save(updatedTask);
            auditGateway.publish(new AuditEntry(updatedTask.id(), "task.completed", clock.instant().toString()));
            return new TaskCompletionResult(true, "OK", updatedTask);
        }
    }
}

File: test files/typescript/pleno_iniciante.ts
type TaskStatus = 'todo' | 'doing' | 'done' | 'blocked';

type Result<T> =
  | { ok: true; data: T }
  | { ok: false; error: string };

interface Task {
  id: number;
  title: string;
  ownerId: number | null;
  status: TaskStatus;
  estimateHours: number;
}

interface TaskRepository {
  findById(id: number): Promise<Task | null>;
  save(task: Task): Promise<Task>;
}

class InMemoryTaskRepository implements TaskRepository {
  private readonly tasks = new Map<number, Task>([
    [1, { id: 1, title: 'Criar backlog', ownerId: 10, status: 'todo', estimateHours: 4 }],
    [2, { id: 2, title: 'Revisar ticket', ownerId: 15, status: 'doing', estimateHours: 2 }],
  ]);

  async findById(id: number): Promise<Task | null> {
    return this.tasks.get(id) ?? null;
  }

  async save(task: Task): Promise<Task> {
    this.tasks.set(task.id, task);
    return task;
  }
}

class TaskWorkflowService {
  constructor(private readonly taskRepository: TaskRepository) {}

  async moveToDone(taskId: number): Promise<Result<Task>> {
    const task = await this.taskRepository.findById(taskId);

    if (!task) {
      return { ok: false, error: 'TASK_NOT_FOUND' };
    }

    if (task.ownerId === null) {
      return { ok: false, error: 'TASK_WITHOUT_OWNER' };
    }

    if (task.estimateHours <= 0) {
      return { ok: false, error: 'INVALID_ESTIMATE' };
    }

    const updatedTask = { ...task, status: 'done' as const };
    await this.taskRepository.save(updatedTask);
    return { ok: true, data: updatedTask };
  }
}

async function run(): Promise<void> {
  const service = new TaskWorkflowService(new InMemoryTaskRepository());
  console.log(await service.moveToDone(1));
}

run();
